### DISCUSSION TITLE
Monitoring Bind security log file while chrooted.

### DISCUSSION THREAD

**Discussion by specializedtom** [fail2ban/fail2ban#3154] [Q&A]
I have been trying to set up security monitoring on my local resolver, as I have been receiving quite a number of failed requests from IPs that are listed on DNSBL's. The setup I am basing my setup on, is via the following blog post : https://www.logcg.com/en/archives/1681.html .

There are however two important differences in my setup : 
1) I am using FirewallD instead of IPTables, and 
2) my installation of Bind is chrooted.

The issue I am having with Fail2Ban is that Fail2Ban is unable to access the security log file when it is enabled on Bind. (Fail2Ban fails to start.) I have tried creating a softlink to the security file outside of the chroot, however Fail2Ban still is unable to read the file. 

Is there a workaround for this issue, or is the only solution to have Bind running non-chrooted?

Suggestions most welcome.

**Comment by sebres** [comment]
> Is there a workaround for this issue, or is the only solution to have Bind running non-chrooted?

A chrooted path is just a path relative "root" you specified for `chroot` of bind... so it must be also found somewhere relative host root.
For instance if your "root" in bind points to `/somepath/bind`, and chrooted path is `/var/log/bind/security.log`, the correct path for outside of chroot would be `/somepath/bind/var/log/bind/security.log`. This is the path you need to specify for fail2ban.
Assumed that it is really chroot and not some container/VM/mount-point... And that fail2ban is running as root (or under the user which is able to access this path).

**Comment by specializedtom** [comment]
When I add the full path to the security.log file (in this case /var/named/chroot/var/named/data/security.log), then restart Fail2Ban, I get the following message : 

`
● fail2ban.service - Fail2Ban Service
     Loaded: loaded (/usr/lib/systemd/system/fail2ban.service; enabled; vendor preset: disabled)
     Active: failed (Result: exit-code) since Tue 2021-11-09 00:11:17 AEST; 5s ago
       Docs: man:fail2ban(1)
    Process: 1976965 ExecStartPre=/bin/mkdir -p /run/fail2ban (code=exited, status=0/SUCCESS)
    Process: 1976966 ExecStart=/usr/bin/fail2ban-server -xf start (code=exited, status=255/EXCEPTION)
   Main PID: 1976966 (code=exited, status=255/EXCEPTION)
        CPU: 193ms

Nov 09 00:11:16 mail.thorptechnology.com.au systemd[1]: Starting Fail2Ban Service...
Nov 09 00:11:16 mail.thorptechnology.com.au systemd[1]: Started Fail2Ban Service.
Nov 09 00:11:17 mail.thorptechnology.com.au fail2ban-server[1976966]: 2021-11-09 00:11:17,170 fail2ban                [1976966]: ERROR   Failed during configuration: Have not found any log file  for named-refused jail
Nov 09 00:11:17 mail.thorptechnology.com.au fail2ban-server[1976966]: 2021-11-09 00:11:17,173 fail2ban                [1976966]: ERROR   Async configuration of server failed
Nov 09 00:11:17 mail.thorptechnology.com.au systemd[1]: fail2ban.service: Main process exited, code=exited, status=255/EXCEPTION
Nov 09 00:11:17 mail.thorptechnology.com.au systemd[1]: fail2ban.service: Failed with result 'exit-code'. `

The content of the [named-refused] portion of my jail.local file is : 
`[named-refused]

port     = domain,953
logpath  = /var/named/chroot/var/named/data/security.log
filter = %(known/filter)s[__line_prefix="\s*\S+\s+named\[\d+\]:"]
action = %(banaction)s[name=%(__name__)s-tcp, port="%(port)s", protocol="tcp", chain="%(chain)s", actname=%(banaction)s-tcp]
         %(banaction)s[name=%(__name__)s-udp, port="%(port)s", protocol="udp", chain="%(chain)s", actname=%(banaction)s-udp]
enabled = true
`
Bind is started using the named-chroot service. I use Fedora as my Linux OS. 

Could it be that Fail2Ban is running as a service that's at issue?

**Reply by sebres** [reply]
Do you see this path (file `/var/named/chroot/var/named/data/security.log`) in a shell?
(Would not it be something like `/var/named/chroot/var/log/security.log` either?)
And which access rights (`ls -la /var/named/chroot/var/named/data/security.log`)?

Anyway if you find it in shell, fail2ban normally would also find it if running as service.

**Reply by specializedtom** [reply]
OK, I have the Fail2Ban service running with the 'named-refused' jail active. It's possibly something to do with SELinux that prevented access to the data directory. I have now changed the security.log file to /var/named/chroot/var/log directory. I also activated the jail on 'Jail.conf' instead of 'Jail.local'. 

However, when I check on the jail using "fail2ban-client status named-refused", Fail2Ban does not recognise the Jail and comes up with the following message : 

`[root@mail filter.d]# fail2ban-client status named-refused

2021-11-09 15:35:56,019 fail2ban                [1988200]: ERROR   NOK: ('named-refused',)

Sorry but the jail 'named-refused' does not exist

[root@mail filter.d]#`

Is there a way I can debug this error?

**Reply by sebres** [reply]
> Is there a way I can debug this error?

Sure. Probably answering the questions I provided above? Do you see this file in shell of root?

As for hidden denial of SELinux (although I don't think so, but...), see https://forums.fedoraforum.org/archive/index.php/t-238861.html for similar issue and follow the instructions (e. g. `setenforce 0`) to debug or investigate deeper.

**Reply by specializedtom** [reply]
I do see the log files in a shell. FYI the access rights for each security.log file is as follows : 

-rw-r--r--. 1 named named system_u:object_r:named_cache_t:s0 569546 Nov  9 15:09 /var/named/chroot/var/named/data/security.log

-rw-r--r--. 1 named named system_u:object_r:named_log_t:s0 149575 Nov 10 14:24 /var/named/chroot/var/log/security.log  (current)

I tried 'setenforce 0' and Fail2Ban was able to access the security.log file in Data directory once both Fail2Ban and Named-Chroot services were restarted, hence my suspicion with SeLinux. 

Currently, I have configured Named to point to the log directory, and have set Fail2Ban to monitor the same (via the named-refused jail). Even with 'setenforce 0', when Fail2Ban starts successfully, Fail2Ban-Client does not recognise the jail (as per previous post). 

Here is my Fail2Ban config : 

Jail.conf
`
[named-refused]

enabled  = true
port     = domain,953
protocol = udp
filter = named-refused
action = %(banaction)s[name=%(__name__)s-tcp, port="%(port)s", protocol="tcp", chain="%(chain)s", actname=%(banaction)s-tcp]
         %(banaction)s[name=%(__name__)s-udp, port="%(port)s", protocol="udp", chain="%(chain)s", actname=%(banaction)s-udp]
logpath  = /var/named/chroot/var/log/security.log
findtime = 1
maxretry = 5

filter.d/named-refused.conf

[Definition]
_daemon=named
__pid_re=(?:\[\d+\])
__daemon_re=\(?%(_daemon)s(?:\(\S+\))?\)?:?
__daemon_combs_re=(?:%(__pid_re)s?:\s+%(__daemon_re)s|%(__daemon_re)s%(__pid_re)s?:)
__line_prefix=(?:\s\S+ %(__daemon_combs_re)s\s+)?

prefregex = ^%(__line_prefix)s(?: error:)?\s*client(?: @\S*)? <HOST>#\S+(?: \([\S.]+\))?: <F-CONTENT>.+</F-CONTENT>\s(?:denied|\(NOTAUTH\))\s*$

failregex = ^(?:view (?:internal|external): )?query(?: \(cache\))?
            ^zone transfer
            ^bad zone transfer request: '\S+/IN': non-authoritative zone

ignoreregex =
`

This does seem to be a configuration issue with Fail2Ban.

**Reply by sebres** [reply]
> This does seem to be a configuration issue with Fail2Ban.

How so? This is surely a permission issue on your system (SELinux or not).
fail2ban-client simply uses glob for a specified `logpath` and if nothing found throw this error.
We can't do something to "fix" it on our side.

**Reply by specializedtom** [reply]
Just did some more digging. It turns out it is a SeLinux issue and it is to do with access to the chrooted "/var/named/chroot" directory. 
Here is the attached sealert from my Journal. 

[fedora@fedora-35-webserver web]$ sudo sealert -l b0900d3c-f676-4c09-9465-73393b8093a4
SELinux is preventing fail2ban-server from search access on the directory chroot.

'*****  Plugin catchall (100. confidence) suggests   **************************'

If you believe that fail2ban-server should be allowed search access on the chroot directory by default.
Then you should report this as a bug.
You can generate a local policy module to allow this access.
Do
allow this access for now by executing:
`# ausearch -c 'fail2ban-server' --raw | audit2allow -M my-fail2banserver `
`# semodule -X 300 -i my-fail2banserver.pp`


Additional Information:
Source Context                system_u:system_r:fail2ban_t:s0
Target Context                system_u:object_r:named_conf_t:s0
Target Objects                chroot [ dir ]
Source                        fail2ban-server
Source Path                   fail2ban-server
Port                          <Unknown>
Host                          fedora-35-webserver
Source RPM Packages           
Target RPM Packages           
SELinux Policy RPM            selinux-policy-targeted-35.5-1.fc35.noarch
Local Policy RPM              fail2ban-selinux-0.11.2-9.fc35.noarch
Selinux Enabled               True
Policy Type                   targeted
Enforcing Mode                Enforcing
Host Name                     fedora-35-webserver
Platform                      Linux fedora-35-webserver 5.14.16-301.fc35.x86_64
                              #1 SMP Wed Nov 3 13:55:42 UTC 2021 x86_64 x86_64
Alert Count                   1
First Seen                    2021-11-11 00:59:50 AEST
Last Seen                     2021-11-11 00:59:50 AEST
Local ID                      b0900d3c-f676-4c09-9465-73393b8093a4

Raw Audit Messages
type=AVC msg=audit(1636556390.944:2458): avc:  denied  { search } for  pid=186085 comm="fail2ban-server" name="chroot" dev="sda1" ino=798991 scontext=system_u:system_r:fail2ban_t:s0 tcontext=system_u:object_r:named_conf_t:s0 tclass=dir permissive=0

Hash: fail2ban-server,fail2ban_t,named_conf_t,dir,search


Upon generating the local policy, Fail2Ban was able to access the log file generated by the Named-Chroot service. 

[fedora@fedora-35-webserver web]$ sudo fail2ban-client status named-refused
Status for the jail: named-refused
|- Filter
|  |- Currently failed:	0
|  |- Total failed:	0
|  `- File list:	/var/named/chroot/var/log/security.log`
`- Actions`
   |- Currently banned:	0
   |- Total banned:	0
   `- Banned IP list:
